ext/mbstring/libmbfl/filters/mbfilter_ucs4.c

```diff
@@ -125,153 +147,146 @@ const struct mbfl_convert_vtbl vtbl_wchar_ucs4le = {
 	NULL,
 };
 
-
 #define CK(statement)	do { if ((statement) < 0) return (-1); } while (0)
 
-/*
- * UCS-4 => wchar
- */
+static int emit_wchar_if_valid(unsigned int w, mbfl_convert_filter *filter)
+{
+	/* FIXME: UCS-4 should actually accept code unit values up to 0x7FFFFFFF
+	 * This is a hack to get around the fact that values of 0x70000000 and up
+	 * are being used to represent illegal characters, which are passed through
+	 * for error handling by a later conversion filter stage
+	 * After we adjust the range of values used for such, change this to 0x7FFFFFFF */
+	if (w < MBFL_WCSGROUP_UCS4MAX) {
+		CK((*filter->output_function)(w, filter->data));
+	} else {
+		CK(mbfl_filt_conv_illegal_output(w, filter));
+	}
+	return 0;
+}
+
 int mbfl_filt_conv_ucs4_wchar(int c, mbfl_convert_filter *filter)
 {
-	int n, endian;
+	/* We start out with the assumption that the string is big-endian
+	 * If we find a little-endian BOM, then we will change that assumption */
+	filter->cache = (filter->cache << 8) | (c & 0xFF);
 
-	endian = filter->status & 0xff00;
-	switch (filter->status & 0xff) {
-	case 0:
-		if (endian) {
-			n = c & 0xff;
-		} else {
-			n = (c & 0xffu) << 24;
-		}
-		filter->cache = n;
-		filter->status++;
-		break;
-	case 1:
-		if (endian) {
-			n = (c & 0xff) << 8;
-		} else {
-			n = (c & 0xff) << 16;
-		}
-		filter->cache |= n;
+	if (filter->status < 3) {
 		filter->status++;
-		break;
-	case 2:
-		if (endian) {
-			n = (c & 0xff) << 16;
-		} else {
-			n = (c & 0xff) << 8;
-		}
-		filter->cache |= n;
-		filter->status++;
-		break;
-	default:
-		if (endian) {
-			n = (c & 0xffu) << 24;
-		} else {
-			n = c & 0xff;
-		}
-		n |= filter->cache;
-		if ((n & 0xffff) == 0 && ((n >> 16) & 0xffff) == 0xfffe) {
-			if (endian) {
-				filter->status = 0;		/* big-endian */
-			} else {
-				filter->status = 0x100;		/* little-endian */
-			}
-			CK((*filter->output_function)(0xfeff, filter->data));
+	} else {
+		if (filter->cache == 0xFFFE0000) {
+			/* If we had the endianness right, the byte-order mark should be 0x0000feff,
+			 * but we got 0xfffe0000, meaning that we are currently using the wrong
+			 * endianness and should switch to little-endian mode */
+			CK((*filter->output_function)(0xFEFF, filter->data));
```

Comment:
So, as we're dropping the BOM for UTF-16, should we be doing that for UTF-32 and UCS-4 as well?

Reply:
I think we should. It seems to make sense.

Will do that.